### DISCUSSION TITLE
COINSTAC Flare First MVP requirements

### DISCUSSION THREAD

**Discussion by spanta28** [trendscenter/coinstac-mint#6] [General]
# COINSTAC MINT MVP Requirements
 
To do items marked in italic
 
**App UI**
 
*_Re-order the panels: 1) consortium details, 2)settings 3)computation details_
* _Design should minimally accommodate a viewport of 800 x 600 , should be able to show 2 columns and sanely display on larger viewports_
* _All applicable views (except Consortium Details) will display fetched data at the moment of loading and should include a refresh button to re-fetch data._
* _Consortium Details will display Live Subscription data at any change/mutation_
     * _Each panel should separately refresh state_
 * _Notifications - Progress bar (Should it be Left to Right)_
* _Icon Based Buttons should have Tooltips_
*_Make sure to tell users clearly who are active or inactive, User should be able to perceive the meaning of check mark on the member icon._
* _Make sure all Icon Based Buttons make sense_

 
**Login**
 
* Tabs
    * Login
        * Username
        * Password 
        * Remember Me
    * Create User
        * Username
        * Password 
        * _Email Address_ 
    * _Username/Password Help_
        * _Forgot Username_
        * _Reset Password_
* _Version Number_
* _App Configuration (Link)_
 
**Consortium List** 
 
* Tabs for Joined/Owned and Other Consortia (All Public for now until Join Hash Feature / or similar implemented)
* Once Joined User Enters Consortium View_
* _Sorted by Date of last Run (most current first). Sorted by last run, because that is the most recently used consortium_
 
    * Each Consortium Item Shows:
    * Name
    * _Date of Creation_
    * _Consortium goal/short description (280 characters)_. Ex: run dpsvm on substance use data
    * Members
    * Join (Enter upon Joining)
 
    * _View (If joined)_
    * _No Leave button (unless member state in list view has live updates)_
    * _A potential solution here: Clicking the 'JOIN'/'Leave' buttons should initiate the refresh page without changing the position of the user on the current consortium_
 
**Consortium Details**
 
* Name
* Short description (EX: Computation that’s being run)
* Indicate when active members mapped their data_
_For showing when active members have mapped the data, put the dir icon below the tick mark on active members to indicate if they mapped the data. Potential solutions:the tick mark could be for active members along with another 'directory/folder' icon on the bottom right on the member icon that shows whether they mapped the data or not... it will be grey/red if they dint, and green if they did thats a solution that I thought of.. or if the icons on the member icon seem too many, then : have a placeholder for "active members":user1,user2 (with a dir icon on their icon). "Inactive users": user3,user4_
*  _(Next release) Join ID_ 
* Computation
* [Adjust Consortium Detail View to be able to open Computation Selection Modal]
Mandatory fields will be :
Title of the computation
Image name
Image download url

> The mandatory fields are what are required to make a computation run in coinstac. This is the solution for a middle ground. The comp
> author will enter the meta info. Into the template. The template will
> have the following keys with possible options. Assuming that this is
> in the database under the meta_info column, we parse the the keys.
> meta_info If the current UI allows a comp author to enter this info.
> so that it gets in the DB, then the comp author/dev can enter the
> info. accurately and parse the keys.
> Indicate Prompt that conveys Importance of the fields below in the creation form

* _GitHub (Project) Repo URL_
* _Image Build Source_
* _Data Type(s) (FSL, NIFTI, etc)_
* _Short Description (text only)_
* Description (Markdown)

**_For Future Consideration_**
Ability to save and store Study Configuration

**Computation**
    * _Ability to minimize panel (perhaps only showing Computation Name)_
 
* _Computation Description should Contain:_
[https://github.com/trendscenter/mint-computation-nvflare-app-boilerplate/blob/main[…]s/computation_publishing/Computation_Publishing_Requirements.md](https://github.com/trendscenter/mint-computation-nvflare-app-boilerplate/blob/main/docs/computation_publishing/Computation_Publishing_Requirements.md)
 
* Members
    * Join / Leave
    * Active 
    * Ready
    * _Owner Editing of Members (Next release)_ 
* Settings (parameters.json)
    * Plain Text Editing
    * JSON Editing
    * _Ability to minimize panel_( Next release)
* Member Data
    * Select 
    * Edit
    * Delete
* Owner Notes
    * Edit
    * Delete
    * _Ability to minimize panel_( Next release) 
* Controls (*Owner only)
    * Start Run*
    * Select Data Directory (if not yet set)
    * _Copy Consortium (Next release_)*
    * _Delete Consortium_*
    * _Download Image_
 
* Runs
* _Display List of Runs (Reveal first 2 Items / Show Scroll if more items)_ 
    * Run Items
        * Date / Time
        * Last Status
        *  Computation Title
        * _Ability to Delete Run(owner only_ 
        * _View Results_
        * _View Details_ [Link to Run instance]
        * Run Id (small font)
        
**Create Consortium**
* Name
* Description
 
**Computation List**
 
* _Sorted Alphabetically_
* Computation Item
    * Name
    * Image Name
 
Extract the following from meta_info or not display some of the following. Info if a solution can’t be reached for MVP.
    * _Data Type(s)? (For eventual filtering/sorting)_
    * _Date of Creation (in system)_
    * _GitHub Repo URL / Link_
    * _Image Build Source_Link_
    * _Short Description (text only)_
    * _Download Image_
* **_Next Release_**
    * _Filter By_
        * _Type_
        * _Data Type_
    * _Search by Name_
 
**Runs List**
* Display only member's runs
* Sorted by date (recent first)
* **_Next Release:_**
    * _Filter by Consortium_
    * _Filter by Computation_
    * _Filter by date_
 
**Run Details**
* Essentially a snapshot of the Consortium and its settings at the time of Run Start
 
**Run Results** 
* A list of all generated files
* iFrame View Window
    * If index.html exists, show by default 
    *  
**Home Page** (Maybe towards public launch)
 
* _Blurb About what the app is and does_
* _Latest Release with Info_
* _Video Demo_
* _Call(s) to action ie. "Create Consortium"_
* _Tally of_ 
    * _Users_
    * _Computations_
    * _Consortia_
* _Link/List of Published Papers_
* _(Eventually) Join Consortium by ID/Hash_

**Future Considerations:**
*How to keep track of computation versions

**Comment by pixelsaurus** [comment]
Here is my time estimate for the above with some description on work categories and roles

**APP UI**

UI Work Estimate: 5 - 8 days

Figma Needs

* A Few Options for Members Panel (0.5 days)
* View for Consortium Details in 800 x 600 to settle on hierarchy of panels (0.5 days)

Notes

* Will work closely with Dylan on Consortium Details (Saving for last bit of work)
* Generally, there will be a re-factor of Frontend to separate API / State / Views

**LOGIN**

UI Work Estimate: 0.5 - 1 day

Backend API Needs: 

* An Email Server
* Add Email as field to Users MongoDB / Apollo

**CONSORTIUM LIST**

UI Work Estimate: 2 - 3 days

Backend API Needs: 

* Need Queries for Joined/Owned Consortia and Otherwise
* Field changes / API Adjustments

**CONSORTIUM DETAILS**

UI Work Estimate: 5 - 8 days

Figma Needs

* Design for Computation Selection Modal

Backend API Needs: 

* Field changes / API Adjustments

**COMPUTATIONS / COMPUTATION LIST**

UI Work Estimate: 2 - 3 days

Backend API Needs: 

* Field changes / API Adjustments

**Reply by dylanmartin** [reply]
I think we should skip password reset via email for now. Details outlined here: https://github.com/trendscenter/coinstac-mint/discussions/8

There are likely minimal or no changes required for the API and data model. We can evaluate as we get to more specific/concrete instances.

**Comment by dylanmartin** [comment]
### Recap from Meeting with Javier - UI Design Patterns & Future Work

- **UI Design Patterns:**
  - Following the `pages` pattern for now, app is structured by distinct pages.
  - Avoid shared components unless there's a clear reason to use them.
  
- **Folder/File Structure:**
  - Folder: `<ComponentName>`
    - Component: `<ComponentName>.tsx`
    - Hook: `use<componentName>.tsx`
    - Context + Hook: `<componentName>Context.tsx` exports `use<ComponentName>Context.tsx`
    - Display: `<ComponentName>Display.tsx`
    - Nested components: `folders [<component1>, <component2>]`
  
- **To-Do:**
  - Add subscription-triggered re-fetch for:
    - Consortium details page
    - Run details page
  - Evaluate and design:
    - Show the latest run in consortium details
    - New runs automatically appear in the latest run component
    - Run element shows
      - live status 
      - link to respective run
    - link to run list page for consortium
  - Refactor `initialize_configs.sh` (small bug)

- **Future Considerations:**
  - UI and backend state: Javier is comfortable adjusting UI to fit backend. 
  - Backend can be changed if needed
  - Focus on keeping things atomic and flexible.

- **Integration Plan:**
  - When integrating with the main codebase, decide if it’s easier to pull from the POC or use the POC as the base.

**Comment by dylanmartin** [comment]
Pages to implement:
- consortium list
- consortium details
- computation list
- computation details
- computation create 
- computation edit
- run list
- run details
- run results
- app settings
- admin
	- set user roles
	- set user password
- login
   - create user
   - reset password

logout button available everywhere

Dylan to implement the create user and reset password in the login page

**Reply by pixelsaurus** [reply]
Also, for ConsortiumDetailsPage/LatestRun I think it'd be good to display the configured Computation Name in the heading as opposed to the Consortium Title

**Comment by dylanmartin** [comment]
Notes from internal dev testing meeting

- Built images need to be cross platform
```
docker buildx build --platform linux/amd64 -t myimage:latest .
docker buildx create --use  
docker buildx build --platform linux/amd64,linux/arm64 -t myimage:latest .
```
- Handle cases where launchNode tries to use an image that doesn't exist
- AppConfig page
	- Allow configuration edits in the page
	- Notify that app restart is required for changes
- nvflare-ssr-csv
	- troubleshoot results not being generated
- Handle "min-clients" problems. Have the run always wait for the active users.
- Results page - bug - links are navigating to home
- Computation page - have the imageDownloadUrl run `docker pull`
- The central server has to have the latest version of the image
- Make it clear to the user that they don't have to do anything else when run status is "ready"
	- Maybe change the status string from "ready" to something else, or display it differently in the UI
